### DISCUSSION TITLE
feature flag/toggle via FFLIB?

### DISCUSSION THREAD

**Discussion by SeanWhalenUKG** [apex-enterprise-patterns/fflib-apex-common#384] [Q&A] [answered]
Is there a standard way to implement feature toggling with the Apex Enterprise Patterns? Some feature-toggling is very granular, even controlling lines within a method. I think that is rightfully out of reach here (correct me if I'm wrong). What I'm looking for is infrastructure to help with [BranchByAbstraction](https://martinfowler.com/bliki/BranchByAbstraction.html) so that the application can opt into different implementations based on flags (probably custom metadata). 

One thought is to add flag logic into the Application.ServiceFactory. In the examples, that code typically just implements a map.get to deliver back the correct implementation class. Would it be reasonable to add code to look at flags that would help govern which of several implementations to instantiate (eg, dev, qa, production, alpha, beta, etc)?

**Comment by daveespo** [comment]
You should give [AT4DX](https://github.com/apex-enterprise-patterns/at4dx) a look -- there is a presentation linked from the wiki that is a good overview of the project

**Reply by SeanWhalenUKG** [reply]
thanks. I have started to look at that and [force-di](https://github.com/apex-enterprise-patterns/force-di) as possible approaches.

**Comment by wimvelzeboer** [comment, marked as answer]
You can also have a look at [fflib-apex-extensions](https://github.com/wimvelzeboer/fflib-apex-extensions), which is kinda lightweight version of AT4DX which also includes a lightweight force-di principle but then focused only on the separations of concerns.

**Reply by SeanWhalenUKG** [reply]
thanks... that looks very helpful. The [Replace](https://github.com/wimvelzeboer/fflib-apex-extensions/wiki/application-factories#fflib_iservicefactory---replacewith) method is exactly what I was looking for. 
And, the documentation in the wiki is excellent!
